test/publish.sh

```diff
@@ -86,6 +87,16 @@ go run main.go build --manifest <(echo "${MANIFEST}")
 
 go run main.go validate --release "${WORK_DIR}/out"
 
+if [[ -z "${COSIGN_KEY}" ]]; then
+  COSIGN_ARGS="--cosignkey ${COSIGN_KEY}"
+fi
+
 if [[ -z "${DRY_RUN:-}" ]]; then
-  go run main.go publish --release "${WORK_DIR}/out" --helmbucket "${HELM_BUCKET}" --gcsbucket "${GCS_BUCKET}" --dockerhub "${DOCKER_HUB}" --dockertags "${VERSION}"
+go run main.go publish --release "${WORK_DIR}/out" \
+  "${COSIGN_ARGS:-}" \
```

Comment:
does this work? I thought this would make a single arg and we want two?

```suggestion
  --cosignkey "${COSIGN_KEY:-}" \
```

May be simpler and give the same behavior

Reply:
While I like John's suggestion, I suspect that it will result in a `flag needs an argument:` error.

Reply:
`--cosignkey "" --githubtoken ...` is different from `--cosignkey --githubtoken ...` I believe. Not 100% sure though

Reply:
Right, the intention for the `${COSIGN_ARGS:-}` placeholder is to hold the entire `--cosignkey gcpkms://path/to/whatever` argument, or skip it with an empty string if no `${COSIGN_KEY}` is present. That means that if a job uses the publish workflow but does not include the `${COSIGN_KEY}` env variable, it will still run, making it easier to migrate without breaking people.

Reply:
I get the intention I just don't think it works?
```
$ x="-l -a"
$ ls "${x}
ls: invalid option -- ' '
Try 'ls --help' for more information.
```

Reply:
I quoted badly -- to do this trick you have to unquote it.
```
$ cat test.sh
───────┬────────────────────────────────────────────────────────────────────
       │ File: test.sh
───────┼────────────────────────────────────────────────────────────────────
   1   │ #!/bin/bash
   2   │ 
   3   │ 
   4   │ 
   5   │ echo "Testing with TESTARG=${TESTARG}"
   6   │ TESTARG="-l -a"
   7   │ ls \
   8   │   ${TESTARG:-} \
   9   │   ./
  10   │ 
  11   │ echo "Testing with TESTARG unset"
  12   │ unset TESTARG
  13   │ ls \
  14   │   ${TESTARG:-} \
  15   │   ./
───────┴────────────────────────────────────────────────────────────────────
$ ./test.sh  
Testing with TESTARG=
total 12
drwxr-x---   2 stewartbutler primarygroup 4096 Sep 21 13:14 .
drwxr-x--- 155 stewartbutler primarygroup 4096 Sep 21 13:13 ..
-rwxr-x---   1 stewartbutler primarygroup  173 Sep 21 13:18 test.sh
Testing with TESTARG unset
test.sh
```

Reply:
> `--cosignkey "" --githubtoken ...` is different from `--cosignkey --githubtoken ...` I believe. Not 100% sure though

I  missed the quotes. I do believe what you suggested John should work fine.

Reply:
Switched to John's idea -- after some testing it looks like the arg parsing is smart enough to deal with empty quotes.